core/slice.go

```diff
@@ -364,89 +316,94 @@ func (sl *Slice) GetPendingHeader() (*types.Header, error) {
+		return types.PendingHeader{Header: newPh, Termini: localPendingHeaderWithTermini.Termini}
+	} else {
+		if domOrigin {
+			newPh = sl.combinePendingHeader(localPendingHeaderWithTermini.Header, domPendingHeader, nodeCtx)
+			return types.PendingHeader{Header: newPh, Termini: localPendingHeaderWithTermini.Termini}
 		}
-	}
-
-	if externPendingHeader.Header.NumberU64() > localPendingHeader.Header.NumberU64() {
-		localPendingHeader.Header = sl.combinePendingHeader(externPendingHeader.Header, localPendingHeader.Header, nodeCtx)
-		localPendingHeader.Termini = externPendingHeader.Termini
-
-		sl.setCurrentPendingHeader(localPendingHeader)
-		sl.phCache[hash] = localPendingHeader
+		return localPendingHeaderWithTermini
 	}
 }
 
 // updatePhCacheFromDom combines the recieved pending header with the pending header stored locally at a given terminus for specified context
 func (sl *Slice) updatePhCacheFromDom(pendingHeader types.PendingHeader, terminiIndex int, indices []int, reorg bool) {
+
 	var localPendingHeader types.PendingHeader
 	hash := pendingHeader.Termini[terminiIndex]
 	localPendingHeader, exists := sl.phCache[hash]
 
-	if !exists { //GENESIS ESCAPE
-		sl.phCache[hash] = pendingHeader
-		sl.pendingHeader = hash
-		return
-	} else {
+	if exists {
 		for _, i := range indices {
 			localPendingHeader.Header = sl.combinePendingHeader(pendingHeader.Header, localPendingHeader.Header, i)
 		}
-		localPendingHeader.Header.SetLocation(pendingHeader.Header.Location())
+		localPendingHeader.Header.SetLocation(common.NodeLocation)
 		sl.phCache[hash] = localPendingHeader
+
+		if reorg {
+			sl.pendingHeader = hash
+		}
 	}
+	log.Warn("no pending header found for", "terminus", hash)
+}
 
-	// Only set the pendingHeader head if the dom did a reorg
+// writePhCache dom writes a given pendingHeaderWithTermini to the cache with the terminus used as the key.
+func (sl *Slice) writeToPhCache(pendingHeaderWithTermini types.PendingHeader) {
+	sl.phCache[pendingHeaderWithTermini.Termini[terminiIndex]] = pendingHeaderWithTermini
+}
+
+// pickPhCacheHead determines if the provided pendingHeader should be selected and returns true if selected
+func (sl *Slice) pickPhCacheHead(reorg bool, externPendingHeaderWithTermini types.PendingHeader) bool {
 	if reorg {
-		sl.pendingHeader = hash
+		sl.pendingHeader = externPendingHeaderWithTermini.Termini[terminiIndex]
+		return true
+	}
+
+	localPendingHeader, exists := sl.phCache[externPendingHeaderWithTermini.Termini[terminiIndex]]
+	if exists && (externPendingHeaderWithTermini.Header.NumberU64() > localPendingHeader.Header.NumberU64()) {
+		sl.setCurrentPendingHeader(externPendingHeaderWithTermini)
+		return true
+	}
+	return false
+}
+
+// setCurrentPendingHeader compares the externPh parent td to the sl.pendingHeader parent td and sets sl.pendingHeader to the exterPh if the td is greater
+func (sl *Slice) setCurrentPendingHeader(externPendingHeader types.PendingHeader) {
```

Comment:
I think this function should be called `updateCurrentPH()`, since it doesn't necessarily set the ph. It may decide not to set it.